src/main/java/org/prebid/server/bidder/openx/OpenxBidder.java

```diff
@@ -0,0 +1,231 @@
+            processedImps = imps.stream().map(this::makeImp).collect(Collectors.toList());
+        } catch (PreBidException e) {
+            errors.add(e.getMessage());
+        }
+
+        final BidRequest ongoingRequest = CollectionUtils.isNotEmpty(processedImps)
+                ? bidRequest.toBuilder().imp(processedImps).ext(makeReqExt(imps)).build()
+                : null;
+
+        return BidRequestWithError.of(ongoingRequest, errors);
+    }
+
+    private Imp makeImp(Imp imp) {
+        final ExtImpOpenx openxImpExt = parseOpenxExt(imp);
+        return imp.toBuilder()
+                .tagid(openxImpExt.getUnit())
+                .bidfloor(openxImpExt.getCustomFloor())
+                .ext(makeImpExt(openxImpExt.getCustomParams()))
+                .build();
+    }
+
+    private ObjectNode makeReqExt(List<Imp> imps) {
+        return Json.mapper.valueToTree(OpenxRequestExt.of(parseOpenxExt(imps.get(0)).getDelDomain(), OPENX_CONFIG));
+    }
+
+    private ExtImpOpenx parseOpenxExt(Imp imp) {
+        final ObjectNode impExt = imp.getExt();
+        final ExtPrebid<?, ExtImpOpenx> impExtPrebid;
+        final ExtImpOpenx impExtOpenx;
+        if (impExt == null) {
+            throw new PreBidException("openx parameters section is missing");
+        }
+
+        try {
+            impExtPrebid = Json.mapper.convertValue(impExt, OPENX_EXT_TYPE_REFERENCE);
+            impExtOpenx = impExtPrebid != null ? impExtPrebid.getBidder() : null;
+            if (impExtOpenx == null) {
+                throw new PreBidException("openx parameters section is missing");
+            }
+        } catch (IllegalArgumentException e) {
+            logger.warn("Error occurred parsing openx parameters", e);
+            throw new PreBidException(e.getMessage(), e);
+        }
+        return impExtOpenx;
+    }
+
+    private ObjectNode makeImpExt(Map<String, String> customParams) {
+        return customParams != null
+                ? Json.mapper.valueToTree(OpenxImpExt.of(Collections.unmodifiableMap(customParams)))
+                : null;
+    }
+
+    private static List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        return bidResponse == null || bidResponse.getSeatbid() == null
+                ? Collections.emptyList()
+                : bidsFromResponse(bidRequest, bidResponse);
+    }
+
+    private static List<BidderBid> bidsFromResponse(BidRequest bidRequest, BidResponse bidResponse) {
+        final Map<String, BidType> impidToBidType = impidToBidType(bidRequest);
+
+        return bidResponse.getSeatbid().stream()
+                .filter(Objects::nonNull)
+                .map(SeatBid::getBid)
+                .filter(Objects::nonNull)
+                .flatMap(Collection::stream)
+                .map(bid -> BidderBid.of(bid, bidType(bid, impidToBidType)))
+                .collect(Collectors.toList());
+    }
+
+    private static Map<String, BidType> impidToBidType(BidRequest bidRequest) {
```

Comment:
bidRequest is used here only to call getImp(), maybe we can change parameters to List<Imp> imps

Reply:
In my opinion the way it is done is more expresive:
impidToBidType(BidRequest bidRequest) sounds to me like do-Something-For-Class,  extract-Something-From-Class

Reply:
Both approaches are used in code-base.